spanner/request_id_header.go

```diff
@@ -0,0 +1,227 @@
+	// Google Engineering has requested that each value be added in Decimal unpadded.
+	// Should we have a standardized endianness: Little Endian or Big Endian?
+	requestID := fmt.Sprintf("%d.%s.%d.%d.%d.%d", xSpannerRequestIDVersion, randIDForProcess, g.id, g.channelID, nthRequest, nthRPC)
+	md := metadata.MD{xSpannerRequestIDHeader: []string{requestID}}
+	return append(priors, grpc.Header(&md))
+}
+
+type requestID string
+
+// augmentErrorWithRequestID introspects error converting it to an *.Error and
+// attaching the subject requestID, unless it is one of the following:
+// * nil
+// * context.Canceled
+// * context.DeadlineExceeded
+// * io.EOF
+// * iterator.Done
+// of which in this case, the original error will be attached as it, since those
+// are sentinel errors used to break sensitive conditions like ending iterations.
+func (r requestID) augmentErrorWithRequestID(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	switch code := status.Code(err); code {
+	case codes.DeadlineExceeded, codes.Canceled:
+		return err
+	}
+
+	switch err {
+	case iterator.Done, io.EOF, context.Canceled, context.DeadlineExceeded:
+		return err
+
+	default:
+		sErr := ToSpannerError(err)
+		if sErr == nil {
+			return err
+		}
+
+		spErr := sErr.(*Error)
+		spErr.RequestID = string(r)
+		return spErr
+	}
+}
+
+func gRPCCallOptionsToRequestID(opts []grpc.CallOption) (reqID requestID, found bool) {
+	for _, opt := range opts {
+		hdrOpt, ok := opt.(grpc.HeaderCallOption)
+		if !ok {
+			continue
+		}
+
+		metadata := hdrOpt.HeaderAddr
+		reqIDs := metadata.Get(xSpannerRequestIDHeader)
+		if len(reqIDs) != 0 && len(reqIDs[0]) != 0 {
+			reqID = requestID(reqIDs[0])
+			found = true
+			break
+		}
+	}
+	return
+}
+
+func (wr *requestIDHeaderInjector) interceptUnary(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	// It is imperative to search for the requestID before the call
+	// because gRPC's internals will consume the headers.
+	reqID, foundRequestID := gRPCCallOptionsToRequestID(opts)
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if !foundRequestID {
+		return err
+	}
+	return reqID.augmentErrorWithRequestID(err)
+}
+
+type requestIDErrWrappingClientStream struct {
+	grpc.ClientStream
+	reqID requestID
+}
+
+func (rew *requestIDErrWrappingClientStream) processFromOutgoingContext(err error) error {
+	if err == nil {
+		return nil
+	}
+	return rew.reqID.augmentErrorWithRequestID(err)
+}
+
+func (rew *requestIDErrWrappingClientStream) SendMsg(msg any) error {
+	err := rew.ClientStream.SendMsg(msg)
+	return rew.processFromOutgoingContext(err)
+}
+
+func (rew *requestIDErrWrappingClientStream) RecvMsg(msg any) error {
+	err := rew.ClientStream.RecvMsg(msg)
+	return rew.processFromOutgoingContext(err)
+}
+
+var _ grpc.ClientStream = (*requestIDErrWrappingClientStream)(nil)
```

Comment:
Compile time type assertions are super useful to catch failing code immediately instead of deep debugging when things fails, they also aid in readability even more.